timers/timers_test.go

```diff
@@ -0,0 +1,417 @@
+		w.AfterFunc(d, func() { wg.Done() })
+	}
+
+	w.Stop()
+	wg.Wait()
+}
+
+func TestPowersOfTwo(t *testing.T) {
+	tests := []struct {
+		in       int
+		out      int
+		exponent int
+	}{
+		{-42, 1, 0},
+		{0, 1, 0},
+		{1, 1, 0},
+		{5, 8, 3},
+		{1000, 1024, 10},
+		{1 << 22, 1 << 22, 22},
+	}
+	for _, tt := range tests {
+		n, exp := nextPowerOfTwo(int64(tt.in))
+		assert.Equal(t, tt.out, int(n), "Unexpected next power of two for input %v", tt.in)
+		assert.Equal(t, tt.exponent, int(exp), "Unexpected exponent for input %v", tt.in)
+	}
+}
+
+func scheduleAndCancel(b *testing.B, w *Wheel) {
+	ds := randomTimeouts(1024, time.Minute)
+	for i := range ds {
+		ds[i] = ds[i] + 24*time.Hour
+	}
+	b.ResetTimer()
+
+	b.RunParallel(func(pb *testing.PB) {
+		i := 0
+		for pb.Next() {
+			w.AfterFunc(ds[i&10], fakeWork).Stop()
+			i++
+		}
+	})
+}
+
+func scheduleAndCancelStd(b *testing.B) {
+	ds := randomTimeouts(1024, time.Minute)
+	for i := range ds {
+		ds[i] = ds[i] + 24*time.Hour
+	}
+	b.ResetTimer()
+
+	b.RunParallel(func(pb *testing.PB) {
+		i := 0
+		for pb.Next() {
+			time.AfterFunc(ds[i&10], fakeWork).Stop()
+			i++
+		}
+	})
+}
+
+func BenchmarkScheduleAndCancelWheelNoHeap(b *testing.B) {
+	w := NewWheel(5*time.Millisecond, 2*time.Minute)
+	defer w.Stop()
+
+	scheduleAndCancel(b, w)
+	b.StopTimer()
+}
+
+func BenchmarkScheduleAndCancelWheelWithHeap(b *testing.B) {
+	w := NewWheel(5*time.Millisecond, 2*time.Minute)
+	defer w.Stop()
+
+	ds := randomTimeouts(10000, time.Minute)
+	for i := range ds {
+		ds[i] = ds[i] + 24*time.Hour
+	}
+	for i := range ds {
+		w.AfterFunc(ds[i], fakeWork)
+	}
+
+	scheduleAndCancel(b, w)
+	b.StopTimer()
+}
+
+func BenchmarkScheduleAndCancelStandardLibraryNoHeap(b *testing.B) {
+	scheduleAndCancelStd(b)
+}
+
+func BenchmarkScheduleAndCancelStandardLibraryWithHeap(b *testing.B) {
+	ds := randomTimeouts(10000, time.Minute)
+	timers := make([]*time.Timer, 10000)
+	for i := range ds {
+		ds[i] = ds[i] + 24*time.Hour
+	}
+	for i := range ds {
+		timers[i] = time.AfterFunc(ds[i], fakeWork)
+	}
+
+	scheduleAndCancelStd(b)
+	b.StopTimer()
+	for _, t := range timers {
+		t.Stop()
+	}
+}
+
+func BenchmarkWheelWorkThread(b *testing.B) {
+	// Note that this benchmark includes 1ms of wait time; attempting to reduce
+	// this by ticking faster is counterproductive, since we start missing
+	// ticks and then need to lock more buckets to catch up.
+	w := NewWheel(time.Millisecond, time.Second)
+	defer w.Stop()
+	var wg sync.WaitGroup
+	b.ResetTimer()
+
+	for i := 0; i < b.N; i++ {
+		wg.Add(1)
+		w.AfterFunc(0, func() { wg.Done() })
+		wg.Wait()
+	}
+}
+
+func BenchmarkWheelTimerExpiry(b *testing.B) {
+	w, c := fakeWheel(time.Millisecond)
+	defer w.Stop()
+	var wg sync.WaitGroup
+	b.ResetTimer()
+
+	for i := 0; i < b.N; i++ {
+		wg.Add(1)
+		w.AfterFunc(time.Millisecond, func() { wg.Done() })
+		c.Add(10 * time.Millisecond)
+		wg.Wait()
+	}
+}
+
+func BenchmarkStandardLibraryTimerExpiry(b *testing.B) {
```

Comment:
Since the standard library doesn't use the mock clock, we should just use the real clock in our real timer wheel, and compare the numbers. It seems odd to advance a mock clock that's not used, but then use real time in this benchmark while the timer wheel benchmark ignores real time.